flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -291,19 +313,22 @@ class ExprLowering {
         });
   }
 
-  // FIXME: replace this
-  mlir::Type peelType(mlir::Type ty, int count) {
-    if (count > 0) {
-      if (auto eleTy = fir::dyn_cast_ptrEleTy(ty))
-        return peelType(eleTy, count - 1);
-      if (auto seqTy = ty.dyn_cast<fir::SequenceType>())
-        return peelType(seqTy.getEleTy(), count - seqTy.getDimension());
-      llvm_unreachable("unhandled type");
+  fir::ExtendedValue genval(Fortran::semantics::SymbolRef sym) {
+    auto loc = getLoc();
+    if (inArrayContext() && !exprCtx.inPreludePhase()) {
+      // `sym` appears in an array expression. This is the code gen phase inside
+      // the loop nest.
+      auto exv = toExtendedValue(symMap.lookupSymbol(sym));
+      if (fir::getBase(exv).getType().isa<fir::SequenceType>()) {
```

Comment:
I, of course, like the type-based solution. 👍 I wish I had done it from the beginning...

I agree that with all the cleanup that you've done, and since it is fresh in your memory, now might be the best time to transition and replace the `!fir.char<KIND>` type with a string type `!fir.char<LEN, KIND>`.